### DISCUSSION TITLE
Screenshot of a webpage to a multi-page PDF

### DISCUSSION THREAD

**Discussion by johnmapeson** [ImageMagick/ImageMagick#7402] [Help] [answered]
### ImageMagick version

7.1.1-33

### Operating system, version and so on

macOS 14.5

### Description

This is a command to convert a screenshot of a web page to a multi-page PDF (US Letter, 8.5 x 11 inches, 2550 x 3300 pixels at 300 pixels per inch, with 1 inch margins), for archival purposes:

```
magick input.png -resize 1950x\> -crop x2700 +repage -bordercolor white -border 300x300 -units PixelsPerInch -density 300 -extent 2550x3300 output.pdf
```

And here is a test screenshot: https://github.com/jsx97/test/blob/main/input.png

The command generally works,

<img width="929" alt="pages" src="https://github.com/ImageMagick/ImageMagick/assets/129062294/26033977-e770-4389-9c91-4880e1144695">

...but as Fred Weinhaus replied me on Stack Overflow, it still has a flaw:

> You added the border all around. If your code adds border on the right and bottom, you could end up with an extra border at the far rows and columns (right and bottom) of your crops. If you do not care about excess on the right and bottom, then that is OK solution.

As far as I know, it is not possible to add a border to the left and top sides only, so I tried to replace `-border` with `-gravity` and `-splice`, but `-gravity` causes the output PDF to be blank. How can this be fixed?

```
magick input.png -resize 1950x\> -crop x2700 +repage -gravity NorthWest -background white -splice 300x300 -units PixelsPerInch -density 300 -extent 2550x3300 output.pdf
```

**Here is a another test image, to make it clear which problem I'm currently trying to solve:** https://github.com/jsx97/test/blob/main/input2.png

If I convert it using the first command, since there is not enought space for bottom margin, a blank page appears:

<img width="702" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/129062294/9378da50-00f0-4c5a-8fbe-a987507d5e6f">

**Comment by fmw42** [comment]
I suggested to you previously in a comment on Stack Overflow to use -splice to add border only to top and left.  See https://imagemagick.org/Usage/crop/#splice.

There is no -geometry in your last command.

**Reply by johnmapeson** [reply]
Sorry, I meant `-gravity`, not `-geometry`. And yes, I discovered your comment in which you mentioned `-splice` even before I tried `-border`. But I still cannot figure out how to add `-splice` to the top and left sides only.

**Reply by fmw42** [reply]
Did you look at the examples on the link that I showed.  In the first set on the far right, the blue is added to the top and left. This would need to be done before cropping your image.

Sorry, I really do not know what you want to do with your command and why you are using -splice 300x300 nor -density 300 nor -extent.   I thought you wanted to simply add a 1 px white border to your input after resizing to the top and left and then add a 1 px border on the bottom and right to each cell. If so then

`magick input.png -resize 1950x\> -gravity northwest -background white -splice 1x1 -crop x2700 +repage -gravity southeast -splice 1x1 output.pdf`

If you need a higher quality output, then add your -density etc. But does that not change the viewing dimensions of the pdf?

**Reply by johnmapeson** [reply]
The command produces blank PDF currently...

**Reply by fmw42** [reply]
Works fine for me.  Check your typing to be sure you have it correctly and no extra spaces where not needed.

But why don't you just put a 1 px border around each cell.

`magick input.png -resize 1950x\> -crop x2700 +repage -bordercolor white -border 1x1  output.pdf`

**Reply by fmw42** [reply]
This is untested as I do not have your input image nor know the dimensions of it.

I think the best way to think about this is what is your raster input at 72 dpi crop size and margin size. Then once you have cropped your input, set the output density to 300 dpi. 

An A4 page at 72 dpi is 612x792.  And your 1 inch border would be 72 px.  So subtract 2*72 from 612 and 792 to get 468x648 so that the bordered frames are 612x792 (including the border).

Now scale those numbers by 300/72 so that the crop size is 1950x2700 and the margin is 300. So

`magick image.png -crop 1950x2700 -bordercolor white -border 300x300 -units pixelsperinch -density 300 output.pdf`

The density of 300 at the end reduces the dimensions from 1950x2700 to 468x648 and with border to 612x792 so that it is an A4 size of 8.5x11 inches.

Modify my command as needed for scaling or padding your input image to be a multiple of the crop size.

**Reply by johnmapeson** [reply]
> Sorry, I really do not know what you want to do with your command

I've added a screenshot to the first post here.

> ...and why you are using `-splice 300x300` nor `-density 300` nor `-extent`.

I use `-density 300` because otherwise the output PDF is blurry. Compare the output of

- `magick input.png -resize 1950x\> -crop x2700 +repage -bordercolor white -border 300x300 -units PixelsPerInch -density 300 -extent 2550x3300 output.pdf` and
- `magick input.png -resize 468x\> -crop x648 +repage -bordercolor white -border 72x72 -extent 612x792 output-nodensity.pdf` (The second command assumes the density is 72, hence 2550x3300 is changed to 612x792, 300x300 to 72x72, etc.)

The input image I'm using here and below is here: https://github.com/jsx97/test/blob/main/input.png. Here, both output PDFs consist of four pages, each page is 8.5 x 11 inches, and their layout is also the same. But the first PDF is sharp, whereas the second is blurry.

I use `-border 300x300` (and I tried to use `-splice` instead of it) to add 1 inch margin.

I use `-extent` to make the last page of the output PDF the same size as others. Compare the output of

- `magick input.png -resize 1950x\> -crop x2700 +repage -bordercolor white -border 300x300 -units PixelsPerInch -density 300 -extent 2550x3300 output.pdf` and
- `magick input.png -resize 1950x\> -crop x2700 +repage -bordercolor white -border 300x300 -units PixelsPerInch -density 300 output-noextent.pdf`

> `magick input.png -resize 1950x\> -gravity northwest -background white -splice 1x1 -crop x2700 +repage -gravity southeast -splice 1x1 output.pdf`

Sorry, but I insist the output PDF is blank. IM 7.1.1-33.

> But why don't you just put a 1 px border around each cell?

I need margins, whereas 1 px border gives almost the same appearance as no margins at all.

> This is untested as I do not have your input image nor know the dimensions of it. <...> `magick image.png -crop 1950x2700 -bordercolor white -border 300x300 -units pixelsperinch -density 300 output.pdf`

I need more time to test it. Currently, the output is something like

<img width="455" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/129062294/715e35e7-a4f1-46ea-bc0d-b2bc93c1e3d4">

Update: In fact, this command is very similar to my first one here on GitHub. Yours:

```
magick image.png -crop 1950x2700 -bordercolor white -border 300x300 -units pixelsperinch -density 300 output.pdf
```

Mine:

```
magick input.png -resize 1950x\> -crop x2700 +repage -bordercolor white -border 300x300 -units PixelsPerInch -density 300 -extent 2550x3300 output.pdf
```

It seems we have simply turned full circle.

**Reply by johnmapeson** [reply]
Here is a another test image, to make it clear which problem I'm currently trying to solve: https://github.com/jsx97/test/blob/main/input2.png

If I convert it using the first command,

```
magick input2.png -resize 1950x\> -crop x2700 +repage -bordercolor white -border 300x300 -units PixelsPerInch -density 300 -extent 2550x3300 output2.pdf
```

sice there is not enought space for bottom margin, a blank page appears:

<img width="702" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/129062294/9378da50-00f0-4c5a-8fbe-a987507d5e6f">

**Comment by fmw42** [comment]
Your empty page comes from a totally white region at the bottom of your input image after resizing. It is the excess of the input image that has no text on it. It then gets filled with white by your -extent. You can see this if you save temporary images right after your crop and compare the last cropped image to the bottom of the resized image.  It is just a strip of white margin at the bottom of the input image.  The only other thing you can do is to look at the last frame after processing and see if it is totally white. Then discard that before doing the pdf of the all the images produced just before the save to pdf.  The only other thing would be to discard any frame that is not a full size. But this risks throwing out any text at the bottom. Alternately, if you find that last image is totally white, then you can write something in it like "Intentionally Left Blank" to let people know.

My modification of your command was (change -bordercolor to -background before -extent) :

```
cd
cd desktop/test
magick input2.png +repage -resize 1950x\> +write t1.png -crop x2700 +repage +write t2.png -bordercolor white -border 300x300 +write t3.png -units PixelsPerInch -density 300 -background white -extent 2550x3300 +write t4.png output2.pdf

```
where I saved intermediate image to review in a folder (test) that just had input2.png

**Reply by johnmapeson** [reply]
Thank you so much. So, my assumption that the blank page was added because of the bottom margin turned out to be wrong, and the blank page was added because of the image itself.

May I ask you the purpose of using `+repage` in the very beginning? And also, can't `-crop 1950x2700` be changed to just `-crop x2700` because there is already `-resize 1950x\>` before it?

**Reply by fmw42** [reply]
Yes, you can change the crop to x2700.  The +repage was added to check that the input PNG does not have any virtual canvas and to remove it if it does.  If it does and you do not add the +repage, you might bet extra totally white results at the end because the virtual canvas could be causing an offset of the actual image.  PNG supports a virtual canvas for example when you crop and when you trim and the result is smaller than you started. It keeps track of the original image size. That is why one has to add +repage after the crop (and after -trim).

**Reply by johnmapeson** [reply]
And one more question. Above, you wrote:

> My modification of your command was (change -bordercolor to -background before -extent) :
>
> `magick input2.png +repage -resize 1950x\> +write t1.png -crop x2700 +repage +write t2.png -bordercolor white -border 300x300 +write t3.png -units PixelsPerInch -density 300 -background white -extent 2550x3300 +write t4.png output2.pdf`

Did you meant that I should change `-bordercolor` to `-background` in your command for some reason?

The other possible interpretation of these words is that they describe what you have done youself, that is, how your commad is different from mine. But since your command uses `-bordercolor`, not `-background`, I don't think this interpretation is correct.

**Reply by fmw42** [reply]
-border needs -bordercolor before it, but -extent needs -background before it.  See my command near the end just before -extent.  Almost all other commands need -background for the color. A few others use -fill. -bordercolor is only used with -border.  The default happens to be white. So your command to use -bordercolor before -extent worked, but only fortunately as it was ignored and the default was used.

**Reply by johnmapeson** [reply, marked as answer]
Thanks a lot. So, to summarize, there can be two versions of this command, with `-border` before or after `-extent`, and with different `-extent` values:

- `magick input.png +repage -resize 1950x\> -crop x2700 +repage -bordercolor white -border 300x300 -background white -extent 2550x3300 -units PixelsPerInch -density 300 output1.pdf`
- `magick input.png +repage -resize 1950x\> -crop x2700 +repage -background white -extent 1950x2700 -bordercolor white -border 300x300 -units PixelsPerInch -density 300 output2.pdf`

Which one makes more sense to you? I ask this because in the current version of your command, you have placed `-extent` after `-border`. But in your previous version (if viewed using GitHub history of edits), you have placed it before `-border`. I try to understand is that simply an arbitrary choice or not.

**Reply by fmw42** [reply]
It was a mistake that was corrected.

**Reply by johnmapeson** [reply]
You mean that the version I have posted here with `-extent` before `-border` may cause some problems? Note that it is not an exact copy of your previous version. It is a slightly changed version of yours, which means it probably doesn't anymore contain mistakes you talk about.

**Reply by fmw42** [reply]
I did not check too thoroughly, but when I looked at the result (and intermediate images), it did not look correct according to your version.  So I fixed your version with a few minor changes.

**Reply by fmw42** [reply]
I must have made a typo before. This command is equivalent to yours

magick input2.png +repage -resize 1950x\> -crop 1950x2700 +repage -background white -extent 1950x2700 -border 300x300  -units pixelsperinch -density 300 output2_fred2.pdf

magick compare -metric rmse output2.pdf[0] output2_fred2.pdf[0] null:
0 (0)

The difference was I used 1950x2700 which is equivalent to your x2700, since the resize made the image width 1950.  Also I used -extent 1950x2700 before changing the density and you used -extent 2550x3300 after padding with -border 300x300 and after changing the density to 300. In your case, 2550/300=8.5 and 3300/300=11.  I also used the correct -background before -extent where you used -bordercolor, which is wrong, but works because the default for -background is white and -bordercolor is ignored for -exent.

My logic for putting -extent 1950x2700 after the crop is because after the crop, the images are supposed to be 1950x2700 so if any (the last one) is not 1950x2700 the -extent will pad to that size. Then I put -border 300x300 to add the  padding at 72 dpi. Then I change the dpi to 300.

**Reply by fmw42** [reply]
I do not understand where you get 1950x2700 in order to get an A4 page.  If we want an A4 8.27x11.69 inch crop at 72 dpi, we get 575x824 and we want to crop so that we allow for a 1 inch border around it at 72 dpi. So 1 inch would be 72 pixels at 72 dpi. Thus we need to subtract 2x72 dpi for the margins. So that is 431x68.

**Reply by fmw42** [reply]
OK. I understand your logic. You worked backwards from 8.5x11 at 300 dpi.  1950/300=6.5 then allowing 300 for 1 inch added to both sides gets to 8.5.  Similarly 2700/300=9 and allowing for 2 inches margin.

**Reply by johnmapeson** [reply]
> I must have made a typo before. This command is equivalent to yours
>
> `magick input2.png +repage -resize 1950x> -crop 1950x2700 +repage -background white -extent 1950x2700 -border 300x300 -units pixelsperinch -density 300 output2_fred2.pdf`

This command, except that there is no `-bordercolor` (which I suppose you missed by accident), is an exact copy of the second I have posted in my "summary":

> So, to summarize, there can be two versions of this command, with -border before or after -extent, and with different -extent values:
>
> - `magick input.png +repage -resize 1950x\> -crop x2700 +repage -bordercolor white -border 300x300 -background white -extent 2550x3300 -units PixelsPerInch -density 300 output1.pdf`
> - `magick input.png +repage -resize 1950x\> -crop x2700 +repage -background white -extent 1950x2700 -bordercolor white -border 300x300 -units PixelsPerInch -density 300 output2.pdf`

I asked you which version of these two you think is better, and judging by your last comment, you prefer the version with `-extent` before `-border`. I completely agree and for the same reason you mentioned:

> (My logic for putting `-extent 1950x2700` after the `-crop` is because after the `-crop`, the images are supposed to be 1950x2700 **so if any (the last one) is not 1950x2700 the `-extent` will pad to that size.** Then I put -border 300x300 to add the padding at 72 dpi. Then I change the dpi to 300.)

In other words, no matter which version is used, with `-extent` before or after `-border`, the output PDF appears the same, but the version with `-extent` before `-border` has more beautiful inner logic.

**Reply by fmw42** [reply]
I don't prefer one over the other. Once I understood that you were working backwards, you version made as much sense as mine.

Yes, in my copy, edit, and paste, the -bordercolor somehow got deleted.

**Reply by johnmapeson** [reply]
> I do not understand where you get 1950x2700 in order to get an A4 page. If we want an A4 8.27x11.69 inch crop at 72 dpi, we get 575x824 and we want to crop so that we allow for a 1 inch border around it at 72 dpi. So 1 inch would be 72 pixels at 72 dpi. Thus we need to subtract 2x72 dpi for the margins. So that is 431x680.

> OK. I understand your logic. You worked backwards from 8.5x11 at 300 dpi. 1950/300=6.5 then allowing 300 for 1 inch added to both sides gets to 8.5. Similarly 2700/300=9 and allowing for 2 inches margin.

By the way, I never talked about A4, which is 210x297 mm and can be rounded to 8.27x11.69 inches. What I talked about is US Letter, which is 8.5x11 inches. They are two different paper sizes.

US Letter at 72 DPI is 612x792 px, and at 300 DPI it is 2550x3300 px. A4 at 72 DPI is around 596x842 px, and at 300 DPI it is around 2481x3508 px (https://pixelsconverter.com/millimeters-to-pixels).

> `magick compare -metric rmse output2.pdf[0] output2_fred2.pdf[0] null:`

Could you explain how this command should be used? For me, it returns

> `zsh: no matches found: output2.pdf[0]`

If I remove `[0]`, it works, but even if the files are identical (and the second file is just a copy of the first), it returns something like

> `12033.3 (0.183617)%`

**Reply by fmw42** [reply]
That is not a reasonable error message for Imagemagick. You must be doing something odd or it is zsh. I always use bash, but that is my preference.

That function is used to get a measure of how close two images are by comparing them pixel to pixel and accumulating a statistic.  The [0] indicates to take only the first page to compare from that file. The measure I chose was rmse -- root mean squared error.  The null: at the end means just report the numerical measure and do not create a difference image. If you put a real output file name, then it will be created and show where the differences are.

```
For novices, see

http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=9620
http://www.imagemagick.org/script/command-line-processing.php
https://imagemagick.org/script/command-line-options.php (and see the sitemap link at the bottom)
http://www.imagemagick.org/Usage/reference.html
http://www.imagemagick.org/Usage/
https://github.com/ImageMagick/usage-markdown
https://imagemagick.org/script/porting.php#cli
https://legacy.imagemagick.org/discourse-server/index.php

```

**Reply by johnmapeson** [reply]
Yes, it turned out in `bash` it works fine. I will figure out which syntax should be used for this in `zsh` and then post it here for reference. (Update: The solution is to enclose filenames in  parenthesis: `'file1.pdf[0]'` and  `'file2.pdf[0]'`.)

There is one more thing I would to understand. [In our discussion at Stack Overflow](https://stackoverflow.com/q/78614536) you suggested to use `-splice`:

> Can you not just add one row and one column to the top and left sides of your input by using `-splice` once for the top and once for the left? Then do your command.

Does using `-splice` have any advantages of using `-border` here? Or did you suggest it simply because I didn't explain the task I tried to solve clear enough?

**Reply by fmw42** [reply]
When I suggested using -splice, I thought you had asked about adding a border just on the bottom and right. -splice is used to add padding on one or more sides, but not generally all sides.  -border is used to add a border to 2 opposite or all 4 sides.
